<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [964]:
import pandas as pd
import psycopg2

In [966]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [967]:
# текст запроса
query_3_1 = f'''select COUNT(*)
FROM public.vacancies
'''

In [968]:
# результат запроса
count_vacancies = pd.read_sql_query(query_3_1, connection)
count_vacancies

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [969]:
# текст запроса
query_3_2 = f'''select COUNT(id)
FROM public.employers
'''

In [970]:
# результат запроса
count_employers = pd.read_sql_query(query_3_2, connection)
count_employers

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [971]:
# текст запроса
query_3_3 = f'''select COUNT(id)
FROM public.areas
'''

In [972]:
# результат запроса
count_areas = pd.read_sql_query(query_3_3, connection)
count_areas

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [973]:
# текст запроса
query_3_4 = '''SELECT COUNT(id)
FROM public.industries
'''

In [974]:
# результат запроса
count_industries = pd.read_sql_query(query_3_4, connection)
count_industries

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    294

***

In [975]:
# выводы по предварительному анализу данных

* В базе данных представлено почти 50 тысяч вакансий от 23.5 тысяч работодателей.
* Количество регионов - 1362.
* Вакансии представлены в 294 сферах деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [976]:
# текст запроса
query_4_1 = '''SELECT a.name area, COUNT(v.id) cnt
FROM public.vacancies v
    JOIN areas a ON v.area_id = a.id
GROUP BY a.name
ORDER BY COUNT(v.id) DESC
'''

In [977]:
# результат запроса
vac_per_area = pd.read_sql_query(query_4_1, connection)
vac_per_area

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


area   cnt
0             Москва  5333
1    Санкт-Петербург  2851
2              Минск  2112
3        Новосибирск  2006
4             Алматы  1892
..               ...   ...
764       Тарко-Сале     1
765    Новоаннинский     1
766            Бирск     1
767           Сасово     1
768             Анна     1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [978]:
# текст запроса
query_4_2 = '''SELECT COUNT(*)
FROM public.vacancies v
WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL''' #условие, что заполнено хотя бы одно поле с зарплатой

In [979]:
# результат запроса
vac_with_salary = pd.read_sql_query(query_4_2, connection)
vac_with_salary

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [980]:
# текст запроса
query_4_3 = '''SELECT round(AVG(salary_from)), round(AVG(salary_to))
FROM public.vacancies v
'''

In [981]:
# результат запроса
avg_salary = pd.read_sql_query(query_4_3, connection)
avg_salary

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


round     round
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [982]:
# текст запроса
query_4_4 = '''SELECT schedule, employment, COUNT(id)
FROM public.vacancies v
GROUP BY 1, 2
ORDER BY 3 DESC
'''

In [983]:
# результат запроса
schedule_employment_cnt = pd.read_sql_query(query_4_4, connection)
schedule_employment_cnt

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [984]:
# текст запроса
query_4_5 = '''SELECT experience
FROM public.vacancies v
GROUP BY experience
ORDER BY COUNT(id)
'''

In [985]:
# результат запроса
experience_cnt = pd.read_sql_query(query_4_5, connection)
experience_cnt

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience
0         Более 6 лет
1           Нет опыта
2       От 3 до 6 лет
3  От 1 года до 3 лет

***

In [986]:
# выводы по детальному анализу вакансий


* Ожидаемо, наибольшее количество вакансий в Москве и крупнейших городах.
* Примерно в половине вакансий указана заработная плата.
* Диапазон зарплат (от среднего минимального значения до среднего максимального) - от 71065 до 110537 рублей в месяц.
* Наиболее востребованные условия работы - на полный рабочий день	при полной занятости. Вакансий с удаленной работой примерно в 5 раз меньше.
* Наибольшее количество вакансий с опытом от 1 года до 3 лет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [987]:
# текст запроса
query_5_1 = '''
--первое место среди работодателей по количеству вакансий
(SELECT e.name
FROM public.vacancies v
    JOIN public.employers e ON v.employer_id = e.id
GROUP BY e.name
ORDER BY COUNT(v.id) DESC 
LIMIT 1)

UNION ALL

--пятое место среди работодателей по количеству вакансий
(SELECT e.name
FROM public.vacancies v
    JOIN public.employers e ON v.employer_id = e.id
GROUP BY e.name
ORDER BY COUNT(v.id) DESC 
OFFSET 4
LIMIT 1)'''

In [988]:
# результат запроса
employers_1_and_5 = pd.read_sql_query(query_5_1, connection)
employers_1_and_5

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name
0         Яндекс
1  Газпром нефть

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [989]:
# текст запроса
query_5_2 = '''SELECT a.name, COUNT(DISTINCT e.id) count_of_employers, COUNT(DISTINCT v.id) count_of_vacancies
FROM public.areas a 
    LEFT JOIN public.employers e ON a.id = e.area
    LEFT JOIN public.vacancies v ON a.id = v.area_id
WHERE v.id IS NULL
GROUP BY a.id
ORDER BY count_of_employers DESC 
LIMIT 1'''

In [990]:
# результат запроса
area_without_vac = pd.read_sql_query(query_5_2, connection)
area_without_vac

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count_of_employers  count_of_vacancies
0  Россия                 410                   0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [991]:
# текст запроса
query_5_3 = '''SELECT e.name, COUNT(DISTINCT v.area_id)
FROM public.vacancies v 
    JOIN public.employers e ON v.employer_id = e.id
GROUP BY e.name
ORDER BY 2 DESC
'''

In [992]:
# результат запроса
employer_count_areas = pd.read_sql_query(query_5_3, connection)
employer_count_areas

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
...                       ...    ...
14761                  UniSol      1
14762            UNISTORY LLC      1
14763                   UNIT6      1
14764     United Distribution      1
14765              UNITEL LLC      1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [993]:
# текст запроса
query_5_4 = '''SELECT COUNT(e.id)
FROM public.employers e 
    LEFT JOIN public.employers_industries ei ON e.id = ei.employer_id
WHERE industry_id IS NULL
'''

In [994]:
# результат запроса
empl_without_indust = pd.read_sql_query(query_5_4, connection)
empl_without_indust

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [995]:
# текст запроса
query_5_5 = '''SELECT e.name
FROM public.employers e 
    JOIN public.employers_industries ei ON e.id = ei.employer_id
GROUP BY e.id
HAVING COUNT(DISTINCT industry_id) = 4 
ORDER BY name
OFFSET 2
LIMIT 1
'''

In [996]:
# результат запроса
empl_4_industries = pd.read_sql_query(query_5_5, connection)
empl_4_industries

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [997]:
# текст запроса
query_5_6 = '''SELECT COUNT(employer_id)
FROM public.employers_industries ei 
    JOIN public.industries i ON ei.industry_id = i.id
WHERE i.name = 'Разработка программного обеспечения'
'''

In [998]:
# результат запроса
employers_develop = pd.read_sql_query(query_5_6, connection)
employers_develop

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [999]:
# код для получения списка городов-милионников

import requests 
from bs4 import BeautifulSoup
from lxml import etree
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
tables = pd.read_html(url)
city_df = tables[1]
city_df['Город'][13] = 'Воронеж'
city_list = tuple(city_df['Город'])
print(city_list)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')


C:\Users\482\AppData\Local\Temp\ipykernel_1116\413336190.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  city_df['Город'][13] = 'Воронеж'


In [1000]:
# текст запроса
query_5_7 = f'''
--выводим список городов-миллионников, в которых представлен Яндекс, и количество вакансий в них
(SELECT a.name, COUNT(v.id)
FROM public.vacancies v
    JOIN public.employers e ON v.employer_id = e.id
    JOIN public.areas a ON v.area_id = a.id
WHERE e.name = 'Яндекс' AND a.name IN {city_list}
GROUP BY a.name)

UNION ALL 

--добавляем результирующую строку с общим количеством вакансий в этих городах
(SELECT 'Total', COUNT(v.id)
FROM public.vacancies v
    JOIN public.employers e ON v.employer_id = e.id
    JOIN public.areas a ON v.area_id = a.id
WHERE e.name = 'Яндекс' AND a.name IN {city_list})
ORDER BY 2
'''

In [1001]:
# результат запроса
yandex_areas = pd.read_sql_query(query_5_7, connection)
yandex_areas

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count
0              Омск     21
1         Челябинск     22
2        Красноярск     23
3         Волгоград     24
4             Пермь     25
5            Казань     25
6    Ростов-на-Дону     25
7               Уфа     26
8            Самара     26
9         Краснодар     30
10          Воронеж     32
11      Новосибирск     35
12  Нижний Новгород     36
13     Екатеринбург     39
14  Санкт-Петербург     42
15           Москва     54
16            Total    485

***

In [1002]:
# выводы по анализу работодателей

* Лидер по числу вакансий на рынке - Яндекс.
* Яндекс охватывает большинство регионов.
* Почти 8500 компаний не указали сферу своей деятельности.
* Разработкой программного обеспечения занимается 3500 компаний.


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [1003]:
# текст запроса
query_6_1 = '''SELECT COUNT(*)
FROM public.vacancies
WHERE name LIKE '%Данн%' OR name LIKE '%Data%' OR name LIKE '%данн%' or name LIKE '%data%' OR name LIKE '%DATA%' OR name LIKE '%ДАНН%'
'''

In [1004]:
# результат запроса
data_vacancies = pd.read_sql_query(query_6_1, connection)
data_vacancies

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [1005]:
# текст запроса
query_6_2 = '''SELECT COUNT(*)
FROM public.vacancies
--условие для начинающего дата-сайентиста
WHERE (name ILIKE '%data scientist%' OR name ILIKE '%data science%' OR name ILIKE '%исследователь данных%' OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
    OR name ILIKE '%machine learning%' OR name ILIKE '%машинн%обучен%')
    AND (name ILIKE '%junior%' OR experience='Нет опыта' OR employment='Стажировка')
'''

In [1006]:
# результат запроса
data_junior_vac = pd.read_sql_query(query_6_2, connection)
data_junior_vac

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [1007]:
# текст запроса
query_6_3 = '''SELECT COUNT(*)
FROM public.vacancies
--условия для отбора вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres
WHERE (name ILIKE '%data scientist%' OR name ILIKE '%data science%' OR name ILIKE '%исследователь данных%' OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
    OR name ILIKE '%machine learning%' OR name ILIKE '%машинн%обучен%')
    AND (key_skills ILIKE '%sql%' OR key_skills ILIKE '%postgres%')
'''

In [1008]:
# результат запроса
data_key_skills = pd.read_sql_query(query_6_3, connection)
data_key_skills

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [1009]:
# текст запроса
query_6_4 = '''SELECT COUNT(*)
FROM public.vacancies
--условия для отбора вакансий для DS, в которых в качестве ключевого навыка указан Python
WHERE (name ILIKE '%data scientist%' OR name ILIKE '%data science%' OR name ILIKE '%исследователь данных%' OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
    OR name ILIKE '%machine learning%' OR name ILIKE '%машинн%обучен%')
    AND key_skills ILIKE '%python%'
'''

In [1010]:
# результат запроса
data_key_skills_python = pd.read_sql_query(query_6_4, connection)
data_key_skills_python

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [1011]:
# текст запроса
query_6_5 = '''
--считаем среднее количество ключевых навыков
SELECT round(AVG(LENGTH(key_skills) - LENGTH(replace(key_skills, CHR(9), '')) + 1), 2) cnt_key_skills
FROM public.vacancies
--отбираем вакансии для DS
WHERE (name ILIKE '%data scientist%' OR name ILIKE '%data science%' OR name ILIKE '%исследователь данных%' OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
    OR name ILIKE '%machine learning%' OR name ILIKE '%машинн%обучен%')
    AND (LENGTH(key_skills) - LENGTH(replace(key_skills, CHR(9), '')) + 1) IS NOT NULL
'''

In [1012]:
# результат запроса
avg_data_key_skills = pd.read_sql_query(query_6_5, connection)
avg_data_key_skills

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt_key_skills
0            6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [1013]:
# текст запроса
query_6_6 = '''SELECT DISTINCT experience exp, round(AVG(case when (salary_from IS NOT NULL AND salary_to IS NOT NULL) then (salary_from + salary_to)/2 else (coalesce(salary_from, salary_to)) end))
FROM public.vacancies
WHERE (name ILIKE '%data scientist%' OR name ILIKE '%data science%' OR name ILIKE '%исследователь данных%' OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
    OR name ILIKE '%machine learning%' OR name ILIKE '%машинн%обучен%')
    AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
GROUP BY exp 
'''

In [1014]:
# результат запроса
avg_salary_data = pd.read_sql_query(query_6_6, connection)
avg_salary_data


c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


exp     round
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

***

In [1015]:
# выводы по предметному анализу

* 3,5% всех вакансий имеют отношение к данным. Это высокий процент при том, что доля ВСЕГО IT в российском ВВП составляет 1% (по данным gazeta.ru).
* Позиция Джуниора очень маловостребована (чуть менее 3% от всех вакансий в Data Science). Молодым специалистам будет очень сложно найти работу.
* Знание языка Python гораздо более востребовано для DataScience специалистов, чем SQL и Postgres.
* В среднем в DataScience вакансиях работодатель указывает 6-7 ключевых навыков. Что логично, ведь это высококвалифицированная сфура деятельности.
* Средняя зарплата в DataScience, судя по данным из нашей таблицы, растёт примерно на 30% в год!

# Общий вывод по проекту

In [1016]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

Дополнительно. Мне, как начинающему дата сайентисту интересно, насколько реально устроиться Джуниору на удалённую работу. Сгруппирую вакансии для Джуниоров по типу рабочего графика.

In [1017]:
# текст запроса
query_extra_1 = '''SELECT schedule, COUNT(id) data_junior_vacancies_cnt
FROM public.vacancies v
WHERE (name ILIKE '%data scientist%' OR name ILIKE '%data science%' OR name ILIKE '%исследователь данных%' OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
    OR name ILIKE '%machine learning%' OR name ILIKE '%машинн%обучен%')
    AND (name ILIKE '%junior%' OR experience='Нет опыта' OR employment='Стажировка')
GROUP BY schedule
ORDER BY COUNT(id) DESC
'''

In [1018]:
# результат запроса
data_junior_vac_schedule = pd.read_sql_query(query_extra_1, connection)
data_junior_vac_schedule

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


schedule  data_junior_vacancies_cnt
0       Полный день                         36
1     Гибкий график                          9
2  Удаленная работа                          6

Аналогично предыдущему запросу посмотрим на предпочтительные типы трудоустройства для джунов.

In [1019]:
# текст запроса
query_extra_2 = '''SELECT employment, COUNT(id) data_junior_vacancies_cnt
FROM public.vacancies v
WHERE (name ILIKE '%data scientist%' OR name ILIKE '%data science%' OR name ILIKE '%исследователь данных%' OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
    OR name ILIKE '%machine learning%' OR name ILIKE '%машинн%обучен%')
    AND (name ILIKE '%junior%' OR experience='Нет опыта' OR employment='Стажировка')
GROUP BY employment
ORDER BY COUNT(id) DESC'''

In [1020]:
# результат запроса
data_junior_vac_employment = pd.read_sql_query(query_extra_2, connection)
data_junior_vac_employment

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


employment  data_junior_vacancies_cnt
0     Полная занятость                         42
1  Частичная занятость                          5
2           Стажировка                          4

Ну и давайте оценим, в каких индустриях больше всего вакансий.

In [1021]:
# текст запроса
query_extra_3 = '''SELECT i.name ind_name, COUNT(v.id) vac_count
FROM public.vacancies v
    JOIN public.employers e ON v.employer_id = e.id 
    JOIN public.employers_industries ei ON e.id = ei.employer_id
    JOIN public.industries i ON ei.industry_id = i.id
GROUP BY 1
ORDER BY 2 DESC
'''

In [1022]:
# результат запроса
popular_industries = pd.read_sql_query(query_extra_3, connection)
popular_industries

c:\Users\482\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


ind_name  vac_count
0                  Разработка программного обеспечения      12499
1    Системная интеграция,  автоматизации технологи...      11034
2    Интернет-компания (поисковики, платежные систе...       6413
3                                                 Банк       2742
4                                      Мобильная связь       1585
..                                                 ...        ...
288                               Спортивная федерация          5
289                                          Химчистка          4
290             Ассоциация в сфере культуры, искусства          2
291                                          Автошкола          1
292                                  Ритуальные услуги          1

[293 rows x 2 columns]

# Выводы
* Мы проанализировали почти 50 тысяч вакансий от 23,5 тысяч работодателей из 294 сфер деятельности.
* Больше всего вакансий в Москве.
* Львиная доля вакансий представлена в сферах: Разработка программного обеспечения (это около 3500 различных компаний), Системная интеграция, автоматизации технологических и бизнес-процессов предприятия, ИТ-консалтинг и в Интернет-Компаниях.
* Диапазон зарплат (от средней минимальной до средней максимальной) - от 71065 до 110537 рублей в месяц.
* Наиболее востребована работа на полный день при полной занятости. Вероятность найти работу удалённо примерно в 5 раз меньше.
* Лидером по количеству вакансий является Яндекс.
* Цифры подтверждают, что работа в IT индустрии, в том числе в Data Science по-прежнему востребована и хорошо оплачиваема.
* 3,5% всех вакансий имеют отношение к данным.
* Средняя зарплата в DataScience, судя по данным из нашей таблицы, растёт примерно на 30% в год!
* Однако устроиться Джуном на работу очень сложно. Джуну лучше рассматривать полную занятость на полный день.